proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -287,130 +251,224 @@ protocol TargetBuildContext {
+        displayName: String,
+        /// The executable to be invoked; should be a tool looked up using
+        /// `tool(named:)`, which may reference either a binary-provided tool
+        /// or a source-built tool.
+        executable: Path,
+        /// Arguments to be passed to the tool. Any paths should be based on
+        /// the paths provided in the target build context.
+        arguments: [String],
+        /// Optional initial working directory of the command.
+        workingDirectory: Path? = nil,
+        /// Any custom settings for the process environment.
+        environment: [String: String] = [:],
+        /// A directory into which the command can write output files that
+        /// should be processed further.
+        outputDirectory: Path
+    )
 }
 
 /// Emits errors, warnings, and remarks to be shown as a result of running
 /// the plugin. If any errors are emitted, the plugin is considered to have
 /// have failed, which will be reported to users during the build.
 protocol DiagnosticsEmitter {
-    func emit(error message: String, file: FilePath? = nil, line: Int? = nil)
-    func emit(warning message: String, file: FilePath? = nil, line: Int? = nil)
-    func emit(remark message: String, file: FilePath? = nil, line: Int? = nil)
+    func emit(error message: String, file: Path? = nil, line: Int? = nil)
+    func emit(warning message: String, file: Path? = nil, line: Int? = nil)
+    func emit(remark message: String, file: Path? = nil, line: Int? = nil)
 }
 
-/// In addition, the API of `FilePath` from SwiftSystem will be available.
+/// Provides information about a list of files. The order is not defined
+/// but is guaranteed to be stable. This allows the implementation to be
+/// more efficient than a static file list.
+protocol FileList: Sequence {
+    func makeIterator() -> FileListIterator
+}
+struct FileListIterator: IteratorProtocol {
+    mutating func next() -> FileInfo?
+}
+
+/// Provides information about a single file in a FileList.
+protocol FileInfo {
+    var path: Path { get }
+    var type: FileType { get }
+}
+
+/// Provides information about a file type. This is expressed as a pro-
+/// tocol so that it can be extended without affecting existing plugins.
+/// It provides additional information about the inputs to the plugin.
+protocol FileType {
+    /// True if the file is a source file handled by the Swift compiler.
+    var isSwiftSource: Bool { get }
+    /// True if the file is a source file handled by the Clang compiler.
+    var isClangSource: Bool { get }
+    /// True if the file is a resource file.
+    var isResource: Bool { get }
+    /// True if the file is one that isn't handled by any built-in rule.
+    var isUnknown: Bool { get }
+}
+
+/// A simple representation of a path in the file system.
+protocol Path {
+    public var filename: String { get }
+    public var stem: String { get }
+    public var `extension`: String? { get }
+    public var parentDirectory { get }
```

Comment:
type?

Reply:
Oops.  Thanks, will fix it.  It's a Path.